lambda-buffers-compiler/test/Test/TypeClassCheck.hs

```diff
@@ -70,3 +101,219 @@ nocycles =
   , mkclass "Monad" ["Applicative"]
   , mkclass "Traversable" ["Foldable", "Functor"]
   ]
+
+---- Solver tests
+
+solveTests :: TestTree
+solveTests =
+  testGroup
+    "Solver tests"
+    [ testCase "1: C [Maybe Int] (completeRules)" $
+        solveTest1 @?= []
+    , testCase "2: D [Maybe Int] (partialRules)" $
+        solveTest2 @?= [dInt]
+    , testCase "3: D [Maybe Int] (complete D, partial C)" $
+        solveTest3 @?= [cInt]
+    , testCase "4: C [[[Bool]]] (completeRules)" $
+        solveTest4 @?= []
+    , testCase "5: C (Either (Either Int Bool) (Either Bool Int)) (completeRules)" $
+        solveTest5 @?= []
+    , testCase "6: C (Either l x) (completeRules)" $
+        solveTest6 @?= [cL, cX]
+    , testCase "7: Sum test (completeRules)" $
+        solveTest7 @?= []
+    , testCase "8: Sum test (partialRules)" $
+        solveTest8 @?= [cBool, cInt]
+    , testCase "9: Rec test (completeRules)" $
+        solveTest9 @?= []
+    , testCase "10: Rec test (partialRules)" $
+        solveTest10 @?= [cBool, cInt]
+    , testCase "11: Prod test (completeRules)" $
+        solveTest11 @?= []
+    , testCase "12: Prod test (partialRules)" $
+        solveTest12 @?= [cBool, cInt]
+    , testCase "13: Dec test (completeRules)" $
+        solveTest13 @?= []
+    , testCase "14: Dec test (partialRules)" $
+        solveTest14 @?= [cBool, cInt]
+    ]
+  where
+    cInt = C _c Int
+    cBool = C _c Bool
+    cL = C _c _l
+    cX = C _c _x
+    dInt = C _d Int
+
+-- hardcoded variables, easier to read than (VarP "blah") everywhere
+_x, _xs, _vars, _name, _l, _body :: Pat
+_x = VarP "x"
+_xs = VarP "xs"
+_vars = VarP "vars"
+_name = VarP "name"
+_l = VarP "l"
+_body = VarP "body"
+
+-- TODO(@gnumonik): Import this from somewhere else when that somewhere else module exists
+mkStructuralRules :: Class -> [Rule]
+mkStructuralRules c =
+  [ C c Nil :<= []
+  , C c (_x :* _xs) :<= [C c _x, C c _xs]
+  , C c (_l := _x) :<= [C c _x]
+  , C c (RecP _xs) :<= [C c _xs]
+  , C c (ProdP _xs) :<= [C c _xs]
+  , C c (SumP _xs) :<= [C c _xs]
+  , C c (DecP _name _vars _body) :<= [C c _body]
+  ]
+
+pattern (:@) :: Pat -> Pat -> Pat
+pattern (:@) p1 p2 = AppP p1 p2
+
+pattern LocalRef :: Text -> Pat
+pattern LocalRef nm = RefP Nil (Name nm)
+
+pattern Maybe :: Pat -> Pat
+pattern Maybe p1 = LocalRef "Maybe" :@ p1
+
+pattern List :: Pat -> Pat
+pattern List p = LocalRef "List" :@ p
+
+pattern Either :: Pat -> Pat -> Pat
+pattern Either l r = (LocalRef "Either" :@ l) :@ r
+
+pattern Int :: Pat
+pattern Int = LocalRef "Int"
+
+pattern Bool :: Pat
+pattern Bool = LocalRef "Bool"
+
+pattern NoConstraints :: Class -> Pat -> Rule
+pattern NoConstraints c p = C c p :<= []
+
+pattern Labeled :: Text -> Pat -> Pat
+pattern Labeled nm p = Name nm := p
+
+userRules1 :: Class -> [Rule]
+userRules1 c =
+  [ NoConstraints c Int
+  , NoConstraints c Bool
```

Comment:
It's nicer with just `C c Int :<= []` imo

Reply:
yeah you're right i got carried away w/ pattern synonyms